openpgp/src/crypto/sdkms.rs

```diff
@@ -839,3 +845,118 @@ fn api_curve_to_sequoia_curve(curve: &ApiCurve) -> Result<SequoiaCurve> {
         _ => Err(Error::msg("cannot convert curve")),
     }
 }
+
+/// Struct with information of hostnames that should not be accessed through a
+/// proxy.
+#[derive(Clone, Debug)]
+
+pub struct NoProxy(Vec<NoProxyEntry>);
+
+impl NoProxy {
+    /// Parse a comma separated list of no proxy hostnames
+    ///
+    /// Valid format of entries:
+    ///     - hostnames
+    ///     - ip
+    ///     - CIDR
+    pub fn parse(no_proxy: &str) -> Self {
+        Self(
+            no_proxy
+                .split(",")
+                .filter_map(|no_proxy| no_proxy.parse().ok())
+                .collect(),
+        )
+    }
+
+    /// Check is a host is in the no proxy list
+    pub fn is_no_proxy(&self, host: &str, port: u16) -> bool {
+        self.0
+            .iter()
+            .any(|no_proxy| Self::is_no_proxy_match(no_proxy, host, port))
+    }
+
+    ///  Check if a host matches a NoProxyEntry
+    fn is_no_proxy_match(no_proxy: &NoProxyEntry, host: &str, port: u16) -> bool {
+        // check for CIDR
+
+        match (&no_proxy.ipnetwork, IpAddr::from_str(host)) {
+            (Some(cidr), Ok(ip)) if cidr.contains(ip) => return true,
+
+            _ => {}
+        }
+
+        // match host fragments
+
+        let matching_hosts = host
+            .split('.')
+            .rev()
+            .take(no_proxy.split_hostname.len())
+            .cmp(no_proxy.split_hostname.iter().rev().map(|val| val.as_ref()))
+            == std::cmp::Ordering::Equal;
+
+        match (matching_hosts, no_proxy.port) {
+            (false, _) => false,
+            (true, None) => true,
+            (true, Some(no_proxy_port)) => port == no_proxy_port,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+
+pub(crate) struct NoProxyEntry {
+    pub ipnetwork: Option<IpNetwork>,
+    pub split_hostname: Vec<String>,
+    pub port: Option<u16>,
+}
+
+impl FromStr for NoProxyEntry {
+    type Err = ();
+
+    fn from_str(s: &str) -> ::std::result::Result<Self, Self::Err> {
+        // split host and port from no_proxy
+
+        let mut no_proxy_splits = s.trim().splitn(2, ":");
+        let no_proxy_host = no_proxy_splits.next().ok_or_else(|| ())?;
+        let no_proxy_port = no_proxy_splits.next()
+            .and_then(|port| port.parse().ok());
+
+        // remove leading dot
+        let no_proxy_host = no_proxy_host.trim_start_matches('.');
+
+        Ok(NoProxyEntry {
+            ipnetwork: IpNetwork::from_str(no_proxy_host).ok(),
+            split_hostname: no_proxy_host.split(".").map(String::from).collect(),
+            port: no_proxy_port,
+        })
+    }
+}
+
+fn decide_proxy_from_env(endpoint: &str) -> Option<Arc::<HyperClient>> {
+    let uri = endpoint.parse::<Uri>().ok()?;
+    let endpoint_host = uri.host()?;
+    let endpoint_port = uri.port().map_or(80, |p| p.as_u16());
+    if let Ok(proxy) = env::var(ENV_HTTP_PROXY) {
```

Comment:
I'd avoid references to environment variables in this crate

Reply:
Now sdkms has its own crate :+1: